### ISSUE TITLE
Detach but no longer active objects fail to set parent

### ISSUE THREAD

**Issue by shadowfinderstudios** [FirstGearGames/FishNet#818]
**General**
Unity version: 6000.0.26f1
Fish-Networking version: 4.5.5R
Discord link: https://discord.com/channels/424284635074134018/851790045534748702/1308093185763049524

**Description**
When the game shuts down there is a possibility of an object that is no longer active but detached from the parent, as a result the SetParent is attempted on it but it isn't active in the hierarchy and causes an exception during shutdown. In addition, NetworkBehavior can attempt to OnDestroy an already null object during this time as well.

**Replication**
1. Create a lot of objects and play your game.
2. Suddenly shut down your game.
3. See the exceptions.

**Expected behavior**
No exceptions on shutdown.

**Code to fix the issue**
In ResetState() in ChildTransformTickSmoother.cs:

`
                    if (_detach && _graphicalObject.gameObject.activeInHierarchy)
                    {
                            _graphicalObject.SetParent(_rootTransform);
                    }
`

And, in NetworkObject.cs change network destroy to:

`
            void NetworkBehaviour_OnDestroy()
            {
                foreach (NetworkBehaviour nb in NetworkBehaviours)
                {
                    if (nb != null) nb.NetworkBehaviour_OnDestroy();
                }
            }
`

**Comment by FirstGearGames** [comment]
Can you elaborate on what parts you're adding these, or what lines you are modifying?

**Comment by shadowfinderstudios** [comment]
Not sure if line numbers will help as much since those can change.

In ResetState() in ChildTransformTickSmoother.cs at line 654, change :
if (_detach) { _graphicalObject.SetParent(_rootTransform); }

To:
if (_detach && _graphicalObject.gameObject.activeInHierarchy) { _graphicalObject.SetParent(_rootTransform); }

And, NetworkObject.cs line 533 in NetworkBehaviour_OnDestroy, change:

void NetworkBehaviour_OnDestroy() { foreach (NetworkBehaviour nb in NetworkBehaviours) { nb.NetworkBehaviour_OnDestroy(); } }

To:
void NetworkBehaviour_OnDestroy() { foreach (NetworkBehaviour nb in NetworkBehaviours) { if (nb != null) nb.NetworkBehaviour_OnDestroy(); } }

Also, I've recently encountered another issue in NetworkObject at line 518-ish where a NetworkObject in the RunTimeParentNetworkBehaviour can be null by the time the program is shut down. The following fix of adding an extra != null for the RuntimeParentNetworkBehaviour.NetworkObject can fix it, as below:

            Observers.Clear();
            if (NetworkBehaviours.Count > 0)
            {
                NetworkBehaviour thisNb = NetworkBehaviours[0];
                if (RuntimeParentNetworkBehaviour != null)
                {
                    if (RuntimeParentNetworkBehaviour.NetworkObject != null)
                        RuntimeParentNetworkBehaviour.NetworkObject.RuntimeChildNetworkBehaviours.Remove(thisNb);
                }
            }

**Comment by FirstGearGames** [comment]
Appreciate it.

**Comment by edunad** [comment]
Having the same issue after updating to the latest fishnet version (pro 4.5.6)
![{41329265-4F05-4747-9604-3087A4628ED4}](https://github.com/user-attachments/assets/c0b5e795-2207-40a7-b764-41a6444f2744)

**Comment by FirstGearGames** [comment]
Yeah, it's not been marked as resolved yet but will be most likely next
release.

**Comment by FirstGearGames** [comment]
I'm changing the detach check to this, which should only try to reattach if not shutting down the game or stopping play mode

```cs
                    if (_detach && !ApplicationState.IsQuitting())
                        _graphicalObject.SetParent(_rootTransform);
```

The runtimeparent one should already be fixed.
Please let me know if issue persist for the graphical one in 4.5.7 release.
If you're still getting the runtimeparent one as well, make a new issue.
